plugins/extraction/ocaml.ml

```diff
@@ -599,16 +599,37 @@ let pp_decl = function
         hov 2 (str "type " ++ ids ++ name ++ def)
     | Dterm (r, a, t) ->
         let def =
-          if is_custom r then str (" = " ^ find_custom r)
+          (* If it it is an external custom, set the def to ': type =  <quote>external_fun_name<quote> '.
+             TODO: I'm not sure, but could we use pp_native_string for quoting the custom name of r?
+          *)
+          if is_external_custom r then str (": ") ++ pp_type false [] t ++ str ( " = \"" ^ find_custom r ^ "\"")
+          (* Otherwise, check if it is a regular custom term. *)
+          else if is_custom r then str (" = " ^ find_custom r)
           else pp_function (empty_env ()) a
         in
         let name = pp_global_name Term r in
-        pp_val name t ++ hov 0 (str "let " ++ name ++ def ++ mt ())
+        (* If it is an external custom, begin the expression with 'external' instead of 'let' *)
+        let expr_begin = if (is_external_custom r) then str "external " else str "let " in
+        (* Make sure that a callback registration is synthesised, if specified for that term. *)
+        let callback_def =
+          try
+            let alias =
+              match (find_callback r) with
+                | Some s -> str s
+                | None   -> name (* No alias specified, use the qualid name. *)
+            in (* TODO: Is using let _ safe here? It does work in OCaml code.*)
+            cut2 () ++ hov 0 ((str "let _ = Callback.register \"") ++ alias ++ (str "\" ") ++ name)
+          with Not_found -> (* No callback registration specified for qualid. *)
+            mt()
+        in pp_val name t ++ hov 0 (expr_begin ++ name ++ def ++ mt ()) ++ callback_def;
+
     | Dfix (rv,defs,typs) ->
         pp_Dfix (rv,defs,typs)
 
 let pp_spec = function
   | Sval (r,_) when is_inline_custom r -> mt ()
+  (* If the function is an external, do not generate the val definition in MLI. *)
+  | Sval (r,_) when is_external_custom r -> mt ()
```

Comment:
This seems wrong. Even if you cannot put a `val` in the `.mli` file, you should still put something there. Otherwise the constant will not be accessible from another extracted module. You should replicate the `external` declaration in the `.mli` file.

Reply:
You are completely right - I should have read the documentation more thoroughly. A val declaration in the .mli is indeed completely valid. Even an external declaration is usable but may lead to side-effects, if I read the documentation correctly. So I adopted that part (I will push my changes to my branch soon)